File: Pufferfish/BaseSession.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data.SQLite;
using OxyPlot;

namespace PufferFish;

public abstract class BaseSession
{
	public const int SESSION_V1 = 1;

	public const int SESSION_V2 = 2;

	public const int maxNumberOfGraphs = 9;

	public const int maxNumberOfSensors = 7;

	public int sessionVersion = 1;

	public int nofgraphs;

	public int nofsensors;

	public ArrayList falls = new ArrayList();

	internal ArrayList gps_data = new ArrayList(5000);

	internal DateTime maxTime;

	internal DateTime minTime;

	public static int gyro_index = 5;

	public static int speed_index = 6;

	internal SQLiteConnection sessionDBConn;

	internal GraphDataSource graphData = new GraphDataSource();

	public uint MinIndex { get; internal set; }

	public uint MaxIndex { get; internal set; }

	public DateTime MaxTime => minTime.AddMilliseconds(MaxIndex - MinIndex);

	public DateTime MinTime => minTime;

	public int getGPSDataCount => gps_data.Count;

	public int getGyroIndex()
	{
		return gyro_index;
	}

	public int getSpeedIndex()
	{
		return speed_index;
	}

	public BaseSession(int nofgraphs, int nofsensors)
	{
		this.nofgraphs = Math.Min(9, nofgraphs);
		this.nofsensors = Math.Min(7, nofsensors);
		MinIndex = 0u;
		MaxIndex = 0u;
	}

	internal Range<uint> GetSessionRange()
	{
		return new Range<uint>
		{
			start = MinIndex,
			end = MaxIndex + 50
		};
	}

	internal abstract void loadData(int minTime, int maxTime);

	public virtual bool isNewCriticalFallFormat()
	{
		return true;
	}

	internal DateTime getTimeForDataIndex(uint selecteDataIndex)
	{
		return minTime.AddMilliseconds(selecteDataIndex - MinIndex);
	}

	internal int getGPSIndex(uint selecteDataIndex)
	{
		for (int i = 0; i < gps_data.Count - 1; i++)
		{
			if (((GPSData)gps_data[i + 1]).index >= selecteDataIndex)
			{
				return i;
			}
		}
		return Math.Max(0, gps_data.Count - 1);
	}

	internal string filename()
	{
		return sessionDBConn.FileName;
	}

	public abstract void CloseDB();

	public abstract void commit();

	public abstract void exportFromTo(string FileName, int minIndex, int maxIndex);

	public IEnumerable<DataPoint> getSensorData(int sensor, int axis)
	{
		return graphData.getSensorData(sensor, axis);
	}

	public IEnumerable<DataPoint> getMainData(int sensor, int axis)
	{
		return graphData.getMainData(sensor, axis);
	}

	public IEnumerable<DataPoint> getSpeedData()
	{
		return graphData.getSpeedData();
	}

	public IEnumerable<DataPoint> getGyroData(int axis)
	{
		return graphData.getGyroData(axis);
	}

	internal GPSData getLastGPSdata()
	{
		try
		{
			return (GPSData)gps_data[gps_data.Count - 1];
		}
		catch (Exception)
		{
			return null;
		}
	}

	internal GPSData getGPSdata(int index)
	{
		return (GPSData)gps_data[index];
	}

	public static string getDBFolder()
	{
		return Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData) + "\\Pufferfish\\db\\";
	}

	public double getMaxLat()
	{
		double max = 0.0;
		foreach (GPSData gpd in gps_data)
		{
			if (gpd.coords.Lat > max)
			{
				max = gpd.coords.Lat;
			}
		}
		return max;
	}

	public double getMaxLng()
	{
		double max = 0.0;
		foreach (GPSData gpd in gps_data)
		{
			if (gpd.coords.Lng > max)
			{
				max = gpd.coords.Lng;
			}
		}
		return max;
	}

	public double getMinLat()
	{
		double min = 0.0;
		foreach (GPSData gpd in gps_data)
		{
			if (gpd.coords.Lat < min)
			{
				min = gpd.coords.Lat;
			}
		}
		return min;
	}

	public double getMinLng()
	{
		double min = 0.0;
		foreach (GPSData gpd in gps_data)
		{
			if (gpd.coords.Lng < min)
			{
				min = gpd.coords.Lng;
			}
		}
		return min;
	}

	public LinkedList<Tag> getTags()
	{
		LinkedList<Tag> ret = new LinkedList<Tag>();
		SQLiteCommand command = new SQLiteCommand("SELECT id, type, description, Timestamp from tags", sessionDBConn);
		SQLiteDataReader reader = command.ExecuteReader();
		try
		{
			while (reader.Read())
			{
				ret.AddLast(new Tag(reader.GetInt64(0), reader.GetString(1), reader.GetString(2), reader.GetDateTime(3)));
			}
		}
		catch (Exception)
		{
		}
		finally
		{
			reader.Close();
		}
		return ret;
	}
}

File: ski-gear-python/session_classes/BaseSession.py
import os
import sqlite3
from abc import ABC, abstractmethod
from datetime import datetime, timedelta
from typing import List, Iterable
from data_classes.GPSData import GPSData
from data_classes.GraphDataSource import DataPoint
from dataclasses import dataclass
import data_classes.GraphDataSource as GraphDataSource

@dataclass
class Tag:
    id: int
    type: str
    description: str
    timestamp: datetime

    # Optional: convenience constructor if you want current timestamp by default
    @classmethod
    def create(cls, id: int, type: str, description: str) -> "Tag":
        return cls(id=id, type=type, description=description, timestamp=datetime.utcnow())

class BaseSession(ABC):
    SESSION_V1 = 1
    SESSION_V2 = 2

    MAX_GRAPHS = 9
    MAX_SENSORS = 7

    gyro_index = 5
    speed_index = 6

    def __init__(self, nofgraphs: int, nofsensors: int):
        self.session_version = 1
        self.nofgraphs = min(self.MAX_GRAPHS, nofgraphs)
        self.nofsensors = min(self.MAX_SENSORS, nofsensors)

        self.falls: List = []
        self.gps_data: List[GPSData] = []

        self.min_time: datetime | None = None
        self.max_time: datetime | None = None

        self.min_index: int = 0
        self.max_index: int = 0

        self.conn: sqlite3.Connection | None = None
        self.graph_data = GraphDataSource.GraphDataSource()

    # -------------------------------
    # Range helpers
    # -------------------------------
    def get_session_range(self) -> tuple[int, int]:
        return (self.min_index, self.max_index + 50)

    # -------------------------------
    # GPS helpers
    # -------------------------------
    @property
    def gps_count(self) -> int:
        return len(self.gps_data)

    def get_time_for_index(self, idx: int) -> datetime:
        if self.min_time is None:
            raise ValueError("min_time not initialized")
        return self.min_time + timedelta(milliseconds=(idx - self.min_index))

    def get_gps_index(self, idx: int) -> int:
        for i in range(len(self.gps_data) - 1):
            if self.gps_data[i+1].index >= idx:
                return i
        return max(0, len(self.gps_data) - 1)

    def get_last_gps(self) -> GPSData | None:
        return self.gps_data[-1] if self.gps_data else None

    def get_db_filename(self) -> str:
        if self.conn is None:
            raise ValueError("Database connection is not initialized")
        return self.conn.execute("PRAGMA database_list").fetchone()[2]

    # -------------------------------
    # Derived classes must implement:
    # -------------------------------
    @abstractmethod
    def load_data(self, min_time: int, max_time: int):
        pass

    @abstractmethod
    def close_db(self):
        pass

    @abstractmethod
    def commit(self):
        pass

    @abstractmethod
    def export_from_to(self, filename: str, min_idx: int, max_idx: int):
        pass

    # -------------------------------
    # Sensor wrappers (GraphDataSource)
    # -------------------------------
    def get_sensor_data(self, sensor: int, axis: int) -> Iterable[DataPoint]:
        return self.graph_data.get_sensor_data(sensor, axis)

    def get_main_data(self, sensor: int, axis: int) -> Iterable[DataPoint]:
        # axis retained for compatibility; underlying method expects only one positional argument
        return self.graph_data.get_main_data(sensor, axis)

    def get_speed_data(self) -> Iterable[DataPoint]:
        return self.graph_data.get_speed_data()

    def get_gyro_data(self, axis: int) -> Iterable[DataPoint]:
        return self.graph_data.get_gyro_data(axis)

    # -------------------------------
    # Tags fetching
    # -------------------------------
    def get_tags(self):
        if self.conn is None:
            raise ValueError("Database connection is not initialized")
        result = []
        cur = self.conn.cursor()
        cur.execute("SELECT id, type, description, Timestamp FROM tags")

        for row in cur.fetchall():
            tag_id, ttype, descr, timestamp = row
            result.append(Tag(tag_id, ttype, descr, datetime.fromisoformat(timestamp)))

        return result

    # -------------------------------
    # Latitude/longitude min/max
    # -------------------------------
    def get_max_lat(self):
        return max((g.coords.Lat for g in self.gps_data), default=0.0)

    def get_min_lat(self):
        return min((g.coords.Lat for g in self.gps_data), default=0.0)

    def get_max_lng(self):
        return max((g.coords.Lng for g in self.gps_data), default=0.0)

    def get_min_lng(self):
        return min((g.coords.Lng for g in self.gps_data), default=0.0)

    @staticmethod
    def get_db_folder() -> str:
        appdata = os.getenv("APPDATA")
        if appdata is None:
            raise EnvironmentError("APPDATA environment variable not set")
        return os.path.join(appdata, "BetterFish", "db")
